File: db4o.net/Db4objects.Db4o.CS/Db4objects.Db4o.CS/Internal/CS/Messages/MReadObject.cs
/* Copyright (C) 2004 - 2008  db4objects Inc.  http://www.db4o.com */

using System;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.CS;
using Db4objects.Db4o.Internal.CS.Messages;

namespace Db4objects.Db4o.Internal.CS.Messages
{
	public sealed class MReadObject : MsgD, IServerSideMessage
	{
		public bool ProcessAtServer()
		{
			StatefulBuffer bytes = null;
			// readWriterByID may fail in certain cases, for instance if
			// and object was deleted by another client
			lock (StreamLock())
			{
				try
				{
					bytes = Stream().ReadWriterByID(Transaction(), _payLoad.ReadInt(), _payLoad.ReadInt
						() == 1);
				}
				catch (Db4oException e)
				{
					WriteException(e);
					return true;
				}
				catch (OutOfMemoryException oome)
				{
					WriteException(new InternalServerError(oome));
					return true;
				}
			}
			if (bytes == null)
			{
				bytes = new StatefulBuffer(Transaction(), 0, 0);
			}
			Write(Msg.ObjectToClient.GetWriter(bytes));
			return true;
		}
	}
}

File: db4oj/core/src/com/db4o/MReadObject.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o;

final class MReadObject extends MsgD {
	final boolean processMessageAtServer(YapSocket sock) {
		YapWriter bytes = null;

		// readObjectByID may fail in certain cases
		// we should look for the cause at some time in the future
		
		YapStream stream = getStream();

		synchronized (stream.i_lock) {
			try {
				bytes = stream.readWriterByID(this.getTransaction(), this.payLoad.readInt());
			} catch (Exception e) {
				bytes = null;
				if (Deploy.debug) {
					System.out.println("MsD.ReadObject:: readObjectByID failed");
				}
			}
		}
		if (bytes == null) {
			bytes = new YapWriter(this.getTransaction(), 0, 0);
		}
		Msg.OBJECT_TO_CLIENT.getWriter(bytes).write(stream, sock);
		return true;
	}
}